Spark 8426

mwws opened the pull request.

wei-mao-intel committed:
add executor blacklist feature

1. Add blackListTracker and blackListStrategy class
2. remove deplicated failedExecutor in TaskSetManager
3. enhance YARN allocator to pass node blacklist info to ResourceManager

wei-mao-intel committed:
some enhancement

1. Add taskId to FailureInfo to be able to contain enough informantion
for backward compatibility
2. replace Seq with Set to speedup contains method and remove duplicate
element by natual
3. remove executors from failureExecutors when
	(1)any task succeed on the executor
	(2)the executor is removed by ExecutorAllocateClient
3. Move IsBlacklist() method into BlacklistTracker to make logic cleaner
4. fit spark code style, eg: lineend, comments start, max length of line

GraceH commented on core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala:
```diff
@@ -0,0 +1,107 @@
+  def getNodeBlacklist(failureExecutors: mutable.HashMap[String, FailureStatus]): Set[String]
+
+  // Default implementation to remove failure executors from HashMap based on given time period.
+  def recoverFailureExecutors(failureExecutors: mutable.HashMap[String, FailureStatus]): Unit = {
```
expiredExecutorsInBlackList

GraceH commented on core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala:
```diff
@@ -0,0 +1,107 @@
+ * The interface to determine executor blacklist and node blacklist.
+ */
+trait BlacklistStrategy {
+  val timeout: Long
```
assigne default value

GraceH replied:
expireTimeInMillisecond

mwws replied:
default value is provided in the apply method in "object BlackStrategy" which is come from SparkConf. Do we still need default value here? or maybe just add a comment to explain it?

GraceH commented on core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala:
```diff
@@ -0,0 +1,107 @@
+ * A node will be in blacklist, if there are more than "nodeFailedThreshold" executors on it
+ * in executor blacklist
+ */
+class ThresholdStrategy (
```
SimpleStrategy?

GraceH commented on core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala:
```diff
@@ -0,0 +1,107 @@
+ * in executor blacklist
+ */
+class ThresholdStrategy (
+    executorFailedThreshold: Int,
```
do you mean maxFailureTaskNumber?

GraceH commented on core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala:
```diff
@@ -0,0 +1,107 @@
+ */
+class ThresholdStrategy (
+    executorFailedThreshold: Int,
+    nodeFailedThreshold: Int,
```
maxBlackExecutorNumber

GraceH commented on core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala:
```diff
@@ -0,0 +1,107 @@
+class ThresholdStrategy (
+    executorFailedThreshold: Int,
+    nodeFailedThreshold: Int,
+    timeout: Long
```
expireTimeInMillisecond

GraceH commented on core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala:
```diff
@@ -0,0 +1,107 @@
+    timeout: Long
+  )extends BlacklistStrategy {
+
+  private def executorBlacklistCandidate(
```
getSelectedExecutorMap

GraceH commented on core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala:
```diff
@@ -0,0 +1,107 @@
+  )extends BlacklistStrategy {
+
+  private def executorBlacklistCandidate(
+      failureExecutors: mutable.HashMap[String, FailureStatus]) = {
```
failedExecutorMap

GraceH commented on core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala:
```diff
@@ -0,0 +1,107 @@
+ * A type of blacklist strategy:
+ * Once task failed  at executor, put the executor and its node into blacklist.
+ */
+class StrictStrategy (timeout: Long) extends BlacklistStrategy {
```
NaiveStrategy?

GraceH commented on core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala:
```diff
@@ -0,0 +1,107 @@
+    sparkConf.get("spark.scheduler.blacklist.strategy", "threshold") match {
+      case "threshold" =>
+        new ThresholdStrategy(
+            sparkConf.getInt("spark.scheduler.blacklist.threshold.executorFailedThreshold", 3),
```
mamAllowedFailureNumber?

GraceH commented on core/src/main/scala/org/apache/spark/scheduler/BlacklistTracker.scala:
```diff
@@ -0,0 +1,121 @@
+  private val strategy = BlacklistStrategy(sparkConf)
+
+  // A daemon thread to expire blacklist executor periodically
+  private val executor = ThreadUtils.newDaemonSingleThreadScheduledExecutor(
```
scheduler ?

GraceH commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -259,7 +248,7 @@ private[spark] class TaskSetManager(
       indexOffset -= 1
       val index = list(indexOffset)
-      if (!executorIsBlacklisted(execId, index)) {
+      if (!blacklistTracker.fold(false)(_.executorIsBlacklisted(execId, sched))) {
```
why we need fold here?  it seems tracker.isBlacklisted(execId, sched) == true/false is enough, right?

GraceH commented on core/src/main/scala/org/apache/spark/scheduler/BlacklistTracker.scala:
```diff
@@ -0,0 +1,121 @@
+
+  def updateFailureExecutors(info: TaskInfo, reason: TaskEndReason) : Unit = synchronized {
+    reason match {
+      case Success => Unit
```
if Task success , to remove that executor from your failure list if exists

GraceH commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -641,7 +618,7 @@ private[spark] class TaskSetManager(
         " because task " + index + " has already completed successfully")
     }
-    failedExecutors.remove(index)
+    blacklistTracker.foreach(_.removeFailureExecutors(Iterable(info.executorId)))
```
blacklistTracker.foreach(_.updateFailureExecutors(info, reason)) or to move this function to higher level

GraceH commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -434,7 +434,12 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
-    doKillExecutors(executorsToKill)
+    val killRequestAcknowledged = doKillExecutors(executorsToKill)
+    if(killRequestAcknowledged) {
+      scheduler.sc.blacklistTracker.foreach(_.removeFailureExecutors(executorsToKill))
```
killed not means that the executor removed.

GraceH replied:
to add that in removeExecutor

wei-mao-intel committed:
fix compile error

fix compile error

GraceH commented on core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala:
```diff
@@ -0,0 +1,107 @@
+ * A type of blacklist strategy:
+ * Once task failed  at executor, put the executor and its node into blacklist.
+ */
+class StrictStrategy (val timeout: Long) extends BlacklistStrategy {
```
it seems strickStrategy is only the example for simpleStrategy by setting number as 1. 

you can remove that or provide an alias for the example.

wei-mao-intel committed:
farther change according to comment

1. rename some varilables and methods
2. correct FailureExecutors logic in CoarseGrainedSchedulerBackend

wei-mao-intel committed:
some change to fit scala style

fit scala style
